### ISSUE TITLE
Fix Flow Layout height not updated properly

### ISSUE THREAD

**Issue by Luxray5474** [Luxray5474/szurubooru_uploader#2]
Summary: 
- The Flow Layout height doesn't get updated after the thumbnails load.

Example:
- Visible as of commit 6add09a 

Reasons:
- The function `FlowLayout.send_thumbnails()` emits a signal for every element in the thumbnail list, one by one. However, it doesn't know if the function is finished or not, so it emits those signals with almost no delay, causing the connected function in the main thread, `MainGUI.add_thumbnail_to_grid()` to pile up with multiple requests. After the `ImportFiles` thread emits it `finished` signal which is connected to `MainGUI.post_load()`, `MainGUI.thumb_layout`'s height gets set to the value of `FlowLayout.total_height`, while only a fraction of the thumbnails have already completed being loaded into the layout. It is not until a `MainGUI.resizeEvent()` that the height gets updated, when most likely all the thumbnails have been loaded already.

Solutions:
- Move the thumbnail loading for-loop from the `ImportFiles` thread into the `MainGUI.add_thumbnail_to_grid()`, and call `MainGUI.post_load()` after it. Using it inside the same thread forces the process to run synchronously, which guarantees that `FlowLayout.total_height` will be the height *after* everytihng has been added.

**Comment by Luxray5474** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Fix Flow Layout height not updated properly](https://trello.com/c/N29aI0El/12-fix-flow-layout-height-not-updated-properly)

**Comment by Luxray5474** [comment]
Update:

While attempting to fix the bug, I have learned a few new things about it, most importantly, the race condition problem is within `FlowLayout`'s internals, because although the process of adding the thumbnails happens in a loop in the main thread, `MainGUI.post_load()` still gets called prematurely. 

Output of most recent save (starting at the point where `MainGUI.add_thumbnails_to_grid()` is called)
```
Adding 67 total elements to the layout...
updating total height 218
updating total height 218
updating total height 218
218
updating total height 218
Loading finished in 20246.081352233887ms.
updating total height 223
updating total height 423
updating total height 428
updating total height 628
updating total height 633
updating total height 833
updating total height 838
updating total height 1038
updating total height 1043
updating total height 1243
updating total height 1248
updating total height 1448
updating total height 1453
updating total height 1653
...
updating total height 6983
```

`...` so that I don't have to put the whole long block in here.

As seen in the output, the message indicating that the loading is done has fired before the final height of the layout has been reached,

Solutions:
- Wait for the widget count to be equal to the number of elements we were adding.

**Comment by Luxray5474** [comment]
**Fixed!**:
- The solution was to create a signal inside `FlowLayout` that is emitted every time `FlowLayout.setGeometry` and thus `FlowLayout.doLayout()` is called. The returned value of `FlowLayout.doLayout` which is the total height the layout needs is emitted. That signal is connected to `MainGUI.update_thumb_layout_wrapper_height()` which sets `MainGUI.thumb_layout_wrapper`'s height to the given value. 
- This solves the problem altogether because the height is updates exactly when it is needed, instead of taking a guess or assuming that everything is already done.
